code/functions.jl

```diff
@@ -0,0 +1,166 @@
+"""
+TODO
+"""
+function generate(x, σ)
+    @assert σ >= 0.0
+    return σ > 0.0 ? rand(TruncatedNormal(x,σ, 0.0, 100.0)) : x
+end
+
+"""
+***Simulations for the host population and specialist parasitoid population dynamics for the first 50 generations***
+
+- `N`: Initial host population size
+- `P`: Initial parasitoid population size
+- `t`: Number of generations
+- `f`: Type of parasite chosen to to the simulations (specialist or generalist)
+- `F`: Finite rate of increase of the host population
+- `D`: Density independent mortality (as a probability of survival)
+- `c`: Number of parasitoids emerging from each host parasitized
+- `a`: Searching efficiency (per capita)
+- `h`: Saturation number of parasitoids
+- `b`: Rate of appoaching h
+- `th`: Handling time (as a proportion of the total time)
+- `m`: Extent of clumping of the parasitoid attacks
+
+Return : `dynamics`: Matrix containing the host and parasite density for the initial populations and for every generation simulated
+"""
+function simulation(N::Float64, P::Float64; t::Int64=50, f=specialist_dyn, F=4.0, D=0.5, c=1.0, a=0.5, h=10.0, b=25.0, th=0.0, m=0.2, D_sd=0.0, a_sd=0.0, h_sd=0.0, c_sd=0.0)
+    # Matrix to store the output
+    dynamics = zeros(Float64, (t+2,7))
+    dynamics[1,2] = N
+    dynamics[1,3] = P
+    dynamics[1,4:7] = [D, a, h, c]
+    # Iterations
+    for current_time in 1:(t+1)
+        current_D=generate(D, D_sd)
+        current_a=generate(a, a_sd)
+        current_h=generate(h, h_sd)
+        current_c=generate(c, c_sd)
+        current_p = @NT(F=F, D=current_D, c=current_c, a=current_a, h=current_h, b=b, th=th, m=m)
+        N_next, P_next = timestep(dynamics[current_time,2], dynamics[current_time,3], current_p; parasite_dyn=f)
+        # Population sizes
+        dynamics[current_time+1,1] = current_time
```

Comment:
`dynamics` is structured an used in a row-major way here.

Reply:
Given the time it takes to run the model, and the fact that replicates are almost never needed, we're OK sacrificing performance if it means keeping the data organised in the way ecologists expect them to. There are a few places throughout where the code is not as optimal as it should, always in the name of clarity (this was done as part of a class).